### ISSUE TITLE
Datetime columns do not work.

### ISSUE THREAD

**Issue by michaelSederlin** [martinfleis/geopandas-view#49]
The current version does not work when there is a datetime or timestamp column in the gdf used. 

```python
import geopandas as gpd
from geopandas_view import view
import folium
from datetime import datetime

nybb = gpd.read_file(gpd.datasets.get_path('nybb'))
nybb["time"] = [datetime.today() for ii in range(len(nybb.index))]

view(nybb)
```

```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
~/miniconda3/envs/python-gis/lib/python3.8/site-packages/IPython/core/formatters.py in __call__(self, obj)
    343             method = get_real_method(obj, self.print_method)
    344             if method is not None:
--> 345                 return method()
    346             return None
    347         else:

~/miniconda3/envs/python-gis/lib/python3.8/site-packages/folium/folium.py in _repr_html_(self, **kwargs)
    297             self._parent = None
    298         else:
--> 299             out = self._parent._repr_html_(**kwargs)
    300         return out
    301 

~/miniconda3/envs/python-gis/lib/python3.8/site-packages/branca/element.py in _repr_html_(self, **kwargs)
    329 
    330         """
--> 331         html = urllib.parse.quote(self.render(**kwargs))
    332         onload = (
    333             'this.contentDocument.open();'

~/miniconda3/envs/python-gis/lib/python3.8/site-packages/branca/element.py in render(self, **kwargs)
    317         """Renders the HTML representation of the element."""
    318         for name, child in self._children.items():
--> 319             child.render(**kwargs)
    320         return self._template.render(this=self, kwargs=kwargs)
    321 

~/miniconda3/envs/python-gis/lib/python3.8/site-packages/folium/folium.py in render(self, **kwargs)
    366             '</style>'), name='map_style')
    367 
--> 368         super(Map, self).render(**kwargs)
    369 
    370     def fit_bounds(self, bounds, padding_top_left=None,

~/miniconda3/envs/python-gis/lib/python3.8/site-packages/folium/elements.py in render(self, **kwargs)
     19             figure.header.add_child(CssLink(url), name=name)
     20 
---> 21         super().render(**kwargs)

~/miniconda3/envs/python-gis/lib/python3.8/site-packages/branca/element.py in render(self, **kwargs)
    641 
    642         for name, element in self._children.items():
--> 643             element.render(**kwargs)

~/miniconda3/envs/python-gis/lib/python3.8/site-packages/folium/features.py in render(self, **kwargs)
    624                 self.highlight_map = mapper.get_highlight_map(
    625                     self.highlight_function)
--> 626         super(GeoJson, self).render()
    627 
    628 

~/miniconda3/envs/python-gis/lib/python3.8/site-packages/branca/element.py in render(self, **kwargs)
    637         script = self._template.module.__dict__.get('script', None)
    638         if script is not None:
--> 639             figure.script.add_child(Element(script(self, kwargs)),
    640                                     name=self.get_name())
    641 

~/miniconda3/envs/python-gis/lib/python3.8/site-packages/jinja2/runtime.py in __call__(self, *args, **kwargs)
    673             )
    674 
--> 675         return self._invoke(arguments, autoescape)
    676 
    677     def _invoke(self, arguments, autoescape):

~/miniconda3/envs/python-gis/lib/python3.8/site-packages/jinja2/runtime.py in _invoke(self, arguments, autoescape)
    677     def _invoke(self, arguments, autoescape):
    678         """This method is being swapped out by the async implementation."""
--> 679         rv = self._func(*arguments)
    680         if autoescape:
    681             rv = Markup(rv)

<template> in macro(l_1_this, l_1_kwargs)

~/miniconda3/envs/python-gis/lib/python3.8/site-packages/jinja2/filters.py in do_tojson(eval_ctx, value, indent)
   1258         options = dict(options)
   1259         options["indent"] = indent
-> 1260     return htmlsafe_json_dumps(value, dumper=dumper, **options)
   1261 
   1262 

~/miniconda3/envs/python-gis/lib/python3.8/site-packages/jinja2/utils.py in htmlsafe_json_dumps(obj, dumper, **kwargs)
    617         dumper = json.dumps
    618     rv = (
--> 619         dumper(obj, **kwargs)
    620         .replace(u"<", u"\\u003c")
    621         .replace(u">", u"\\u003e")

~/miniconda3/envs/python-gis/lib/python3.8/json/__init__.py in dumps(obj, skipkeys, ensure_ascii, check_circular, allow_nan, cls, indent, separators, default, sort_keys, **kw)
    232     if cls is None:
    233         cls = JSONEncoder
--> 234     return cls(
    235         skipkeys=skipkeys, ensure_ascii=ensure_ascii,
    236         check_circular=check_circular, allow_nan=allow_nan, indent=indent,

~/miniconda3/envs/python-gis/lib/python3.8/json/encoder.py in encode(self, o)
    197         # exceptions aren't as detailed.  The list call should be roughly
    198         # equivalent to the PySequence_Fast that ''.join() would do.
--> 199         chunks = self.iterencode(o, _one_shot=True)
    200         if not isinstance(chunks, (list, tuple)):
    201             chunks = list(chunks)

~/miniconda3/envs/python-gis/lib/python3.8/json/encoder.py in iterencode(self, o, _one_shot)
    255                 self.key_separator, self.item_separator, self.sort_keys,
    256                 self.skipkeys, _one_shot)
--> 257         return _iterencode(o, 0)
    258 
    259 def _make_iterencode(markers, _default, _encoder, _indent, _floatstr,

~/miniconda3/envs/python-gis/lib/python3.8/json/encoder.py in default(self, o)
    177 
    178         """
--> 179         raise TypeError(f'Object of type {o.__class__.__name__} '
    180                         f'is not JSON serializable')
    181 

TypeError: Object of type Timestamp is not JSON serializable

<folium.folium.Map at 0x7f12080e0040>
```

**Comment by martinfleis** [comment]
Good catch, thanks!

There are two issues linked to datetime dtype. One is this one above, which we could resolve by converting the column to a string before passing it to folium.

The second issue happens if you want to use it as `column` to plot a choropleth but that is not supported even in static plotting at the moment.

**Comment by jorisvandenbossche** [comment]
@martinfleis do we have any control over how the data is converted to JSON? Or could we do that on our side?  (in addition to dealing with non-builtin types, doing it ourselves could potentially also be faster. Although given this is for "small" data anyway, that might not matter much)

**Comment by martinfleis** [comment]
> do we have any control over how the data is converted to JSON? Or could we do that on our side?

We pass data as `__geo_interface__`, so we do that on our side. `view` works on a copy of `gdf` so we can just call `astype(str)`.

> Although given this is for "small" data anyway

I tested 140k building footprints and had a pretty smooth behaviour with `prefer_canvas=True` option :), so not so "small" necessarily.

**Comment by jorisvandenbossche** [comment]
> We pass data as `__geo_interface__`, so we do that on our side. `view` works on a copy of `gdf` so we can just call `astype(str)`.

Ah, so it's actually a "bug" on the geopandas side, as you get the same error with `to_json`:

```
In [29]: nybb = gpd.read_file(gpd.datasets.get_path('nybb'))
    ...: nybb["time"] = [datetime.today() for ii in range(len(nybb.index))]

In [30]: nybb.to_json()
...
TypeError: Object of type Timestamp is not JSON serializable

```

Folium accepts a string as raw JSON as well, so (long term) we could also fix the timestamp issue in general and pass `gdf.to_json()` instead of `gdf.__geo_interface__`.

**Comment by martinfleis** [comment]
I'll move this issue to GeoPandas and switch to `to_json` here directly. Once `gdf.to_json()` will be fixed, this will get resolved automatically and we don't have to deal with it twice.

**Comment by michaelSederlin** [comment]
Hi, 

It seems the same problem occurs when there is a Point geometry column in the data set that is not the set geometry column. 

Figured it's not worth opening a new issue for. 

Best, 
Michael

```
--------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
/lib/python3.8/site-packages/IPython/core/formatters.py in __call__(self, obj)
    343             method = get_real_method(obj, self.print_method)
    344             if method is not None:
--> 345                 return method()
    346             return None
    347         else:

/lib/python3.8/site-packages/folium/folium.py in _repr_html_(self, **kwargs)
    297             self._parent = None
    298         else:
--> 299             out = self._parent._repr_html_(**kwargs)
    300         return out
    301 

/lib/python3.8/site-packages/branca/element.py in _repr_html_(self, **kwargs)
    329 
    330         """
--> 331         html = urllib.parse.quote(self.render(**kwargs))
    332         onload = (
    333             'this.contentDocument.open();'

/lib/python3.8/site-packages/branca/element.py in render(self, **kwargs)
    317         """Renders the HTML representation of the element."""
    318         for name, child in self._children.items():
--> 319             child.render(**kwargs)
    320         return self._template.render(this=self, kwargs=kwargs)
    321 

/lib/python3.8/site-packages/folium/folium.py in render(self, **kwargs)
    366             '</style>'), name='map_style')
    367 
--> 368         super(Map, self).render(**kwargs)
    369 
    370     def fit_bounds(self, bounds, padding_top_left=None,

/lib/python3.8/site-packages/folium/elements.py in render(self, **kwargs)
     19             figure.header.add_child(CssLink(url), name=name)
     20 
---> 21         super().render(**kwargs)

/lib/python3.8/site-packages/branca/element.py in render(self, **kwargs)
    641 
    642         for name, element in self._children.items():
--> 643             element.render(**kwargs)

/lib/python3.8/site-packages/folium/features.py in render(self, **kwargs)
    624                 self.highlight_map = mapper.get_highlight_map(
    625                     self.highlight_function)
--> 626         super(GeoJson, self).render()
    627 
    628 

/lib/python3.8/site-packages/branca/element.py in render(self, **kwargs)
    637         script = self._template.module.__dict__.get('script', None)
    638         if script is not None:
--> 639             figure.script.add_child(Element(script(self, kwargs)),
    640                                     name=self.get_name())
    641 

/lib/python3.8/site-packages/jinja2/runtime.py in __call__(self, *args, **kwargs)
    673             )
    674 
--> 675         return self._invoke(arguments, autoescape)
    676 
    677     def _invoke(self, arguments, autoescape):

/lib/python3.8/site-packages/jinja2/runtime.py in _invoke(self, arguments, autoescape)
    677     def _invoke(self, arguments, autoescape):
    678         """This method is being swapped out by the async implementation."""
--> 679         rv = self._func(*arguments)
    680         if autoescape:
    681             rv = Markup(rv)

<template> in macro(l_1_this, l_1_kwargs)

/lib/python3.8/site-packages/jinja2/filters.py in do_tojson(eval_ctx, value, indent)
   1258         options = dict(options)
   1259         options["indent"] = indent
-> 1260     return htmlsafe_json_dumps(value, dumper=dumper, **options)
   1261 
   1262 

/lib/python3.8/site-packages/jinja2/utils.py in htmlsafe_json_dumps(obj, dumper, **kwargs)
    617         dumper = json.dumps
    618     rv = (
--> 619         dumper(obj, **kwargs)
    620         .replace(u"<", u"\\u003c")
    621         .replace(u">", u"\\u003e")

/lib/python3.8/json/__init__.py in dumps(obj, skipkeys, ensure_ascii, check_circular, allow_nan, cls, indent, separators, default, sort_keys, **kw)
    232     if cls is None:
    233         cls = JSONEncoder
--> 234     return cls(
    235         skipkeys=skipkeys, ensure_ascii=ensure_ascii,
    236         check_circular=check_circular, allow_nan=allow_nan, indent=indent,

/lib/python3.8/json/encoder.py in encode(self, o)
    197         # exceptions aren't as detailed.  The list call should be roughly
    198         # equivalent to the PySequence_Fast that ''.join() would do.
--> 199         chunks = self.iterencode(o, _one_shot=True)
    200         if not isinstance(chunks, (list, tuple)):
    201             chunks = list(chunks)

~/miniconda3/envs/METRIC/lib/python3.8/json/encoder.py in iterencode(self, o, _one_shot)
    255                 self.key_separator, self.item_separator, self.sort_keys,
    256                 self.skipkeys, _one_shot)
--> 257         return _iterencode(o, 0)
    258 
    259 def _make_iterencode(markers, _default, _encoder, _indent, _floatstr,

/lib/python3.8/json/encoder.py in default(self, o)
    177 
    178         """
--> 179         raise TypeError(f'Object of type {o.__class__.__name__} '
    180                         f'is not JSON serializable')
    181 

TypeError: Object of type Point is not JSON serializable

<folium.folium.Map at 0x7fd4d0423340>
```

**Comment by martinfleis** [comment]
Good catch. We'll need to add support of additional dtypes to geopandas json tooling. There will likely be other causing issues.

**Comment by martinfleis** [comment]
We'll try to ensure that this gets fixed by geopandas/geopandas#1906.
